### DISCUSSION TITLE
How to simulate changes in crazyflie firmware?

### DISCUSSION THREAD

**Discussion by zafer4481** [USC-ACTLab/crazyswarm#639] [Q&A]
I installed crazyswarm on linux for simulation only, which works totally fine.

I wish to change the firmware (controller.c stabilizer.c etc) and simulate how my own implemted controller would work. To see if I can simulate the changes in firmware I set the powerdistributions to 0 and ran the hello_world.py --sim script hoping the drone wouldn't move at all, but it did. So it seems like I was not succesful in changing the firmware and simulating the effect.

Probably I am missing some steps, any idea's? Is this even possible what I am trying to do?

**Comment by knmcguire** [comment]
Hi! The --sim does not simulate the controllers and dynamics of the crazyflie, only the reference trajectory that you generated if I understand correctly. So if you want to test out controllers you would need to use another simulator. This would be either [CrazyS](https://github.com/gsilano/CrazyS) or you can check out the recent developments of the [crazyflie-simulation](https://github.com/bitcraze/crazyflie-simulation) repo. It is still experimental, but it flies in Webots already with pythonbindings of the crazyflie onboard controllers! Perhaps that would be good to try it out and see if this works for you?

**Reply by zafer4481** [reply]
I managed to make it work with cf_sim: https://github.com/wuwushrek/sim_cf

I can make changes in crazyflie-firmware, build that code and simulate it in gazebo. The documentation seems a bit outdated, but I installed it in ubuntu 20.04 and it seems to work fine. Im quite happy with it, thank you very much!

CrazyS or crazyflie-simulation don't seem to make use of the exact same crazyflie-firmware folder which is uploaded on the crazyflie hardware (which is what i was looking for). But thanks anyway for informing me about those options as well!

**Reply by knmcguire** [reply]
Ah great, yes indeed sim_cf uses HITL. Unfortunately the repository is no longer actively maintained I heard so I hope it will still continue to work for your purposes!

**Comment by zafer4481** [comment]
I am only intending to change the controller, the algorithm between reading sensor values to actuating the motors. Because I only have access to the crazyflies at my university, I intend to use this simulation to write/test code at home. Every now and then I will test code on hardware at university. I will let you know how successful this will be.

________________________________
